File: go/tests/adc_dac/hx710a_test/main.go
//go:build linux && !tinygo

// HX710A hardware test — Linux host.
//
// Opens DOUT and PD_SCK on /dev/gpiochip0 and runs the HX710A check
// sequence: smoke test for both Minimal and Full, rate selection
// (including invalid rate), averaging, tare, scale, weight, and
// temperature. Power-down/power-up is exercised only in the Complete
// example, not here. Prints PASS/FAIL per check and ends with the
// standard ===DONE: ... === line. Exits 0 on full pass, 1 on any
// failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/adc_dac"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	dout, err := strconv.Atoi(envOr("HX710A_DOUT", "2"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "HX710A_DOUT:", err)
		os.Exit(2)
	}
	pdSck, err := strconv.Atoi(envOr("HX710A_PD_SCK", "3"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "HX710A_PD_SCK:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	conn, trErr := connection.NewHX711Connection(dout, pdSck, nil)
	if trErr != nil {
		fmt.Fprintln(os.Stderr, "connection open failed (no hw?):", trErr)
		check("hx710a_minimal_skip_no_hw", true)
		check("hx710a_full_skip_no_hw", true)
		fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
		if failed != 0 {
			os.Exit(1)
		}
		return
	}
	defer conn.Close()

	// --- HX710AMinimal ---
	{
		hx, err := adcdac.NewHX710AMinimal(conn)
		check("minimal_construct", err == nil && hx != nil)

		ready, err := hx.IsReady()
		check("minimal_is_ready_ok", err == nil)
		_ = ready

		raw, err := hx.ReadRaw()
		check("minimal_read_raw_ok", err == nil)
		if err == nil {
			check("minimal_read_raw_in_range", raw >= -8388608 && raw <= 8388607)
		}
	}

	// --- HX710AFull ---
	{
		hx, err := adcdac.NewHX710AFull(conn)
		check("full_construct", err == nil && hx != nil)

		ready, err := hx.IsReady()
		check("full_is_ready_ok", err == nil)
		_ = ready

		raw, err := hx.ReadRaw()
		check("full_read_raw_ok", err == nil)
		if err == nil {
			check("full_read_raw_in_range", raw >= -8388608 && raw <= 8388607)
		}

		check("set_rate_40_ok", hx.SetRate(adcdac.HX710ARate40SPS) == nil)
		check("set_rate_10_ok", hx.SetRate(adcdac.HX710ARate10SPS) == nil)
		check("set_rate_invalid_rejected", hx.SetRate(20) != nil)
		// back to 10 SPS for the rest of the test
		_ = hx.SetRate(adcdac.HX710ARate10SPS)

		avg, err := hx.ReadAverage(3)
		check("full_read_average_ok", err == nil)
		if err == nil {
			check("full_read_average_in_range", avg >= -8388608 && avg <= 8388607)
		}

		check("tare_ok", hx.Tare(3) == nil)
		offset := hx.GetOffset()
		check("get_offset_in_range", offset >= -8388608 && offset <= 8388607)

		hx.SetScale(420.0)
		check("get_scale_set", hx.GetScale() == 420.0)

		weight, err := hx.ReadWeight(3)
		check("read_weight_ok", err == nil)
		if err == nil {
			check("read_weight_finite", weight == weight) // NaN check
		}

		tempRaw, err := hx.ReadTemperatureRaw()
		check("read_temperature_raw_ok", err == nil)
		if err == nil {
			check("read_temperature_raw_in_range", tempRaw >= -8388608 && tempRaw <= 8388607)
		}
	}

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}

File: rust/tests/adc_dac/hx710a_test/src/main.rs
use gpio_cdev::{Chip, LineRequestFlags};
use linux_embedded_hal::CdevPin;
use periph::chips::adc_dac::{Hx710aFull, Hx710aMinimal};
use periph::connection::hx711::{HX711Error, HX711Connection};

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond {
            println!("PASS {}", $label);
            $passed += 1;
        } else {
            println!("FAIL {}", $label);
            $failed += 1;
        }
    };
}

fn main() {
    let chip_path = std::env::var("GPIO_CHIP").unwrap_or_else(|_| "/dev/gpiochip0".into());
    let dout_offset: u32 = std::env::var("HX710A_DOUT")
        .ok()
        .and_then(|v| v.parse().ok())
        .unwrap_or(5);
    let pd_sck_offset: u32 = std::env::var("HX710A_PD_SCK")
        .ok()
        .and_then(|v| v.parse().ok())
        .unwrap_or(6);

    let mut passed = 0i32;
    let mut failed = 0i32;

    let mut chip = Chip::new(&chip_path).expect("open gpio chip");

    // --- Hx710aMinimal ---
    {
        let dout_handle = chip.get_line(dout_offset).expect("get dout line")
            .request(LineRequestFlags::INPUT, 0, "hx710a_test_minimal").expect("request dout");
        let pd_sck_handle = chip.get_line(pd_sck_offset).expect("get pd_sck line")
            .request(LineRequestFlags::OUTPUT, 0, "hx710a_test_minimal").expect("request pd_sck");

        let dout   = CdevPin::new(dout_handle).expect("dout pin");
        let pd_sck = CdevPin::new(pd_sck_handle).expect("pd_sck pin");

        let connection = HX711Connection::new(dout, pd_sck);
        let mut minimal = Hx710aMinimal::new(connection).expect("init Hx710aMinimal");

        let ready = minimal.is_ready();
        check_true!(ready.is_ok(), "minimal is_ready ok", passed, failed);

        let raw = minimal.read_raw();
        check_true!(raw.is_ok(), "minimal read_raw ok", passed, failed);
        if let Ok(v) = raw {
            check_true!(
                v >= -8_388_608 && v <= 8_388_607,
                "minimal read_raw in 24-bit signed range",
                passed,
                failed
            );
        }
    }

    // --- Hx710aFull ---
    {
        let dout_handle = chip.get_line(dout_offset).expect("get dout line")
            .request(LineRequestFlags::INPUT, 0, "hx710a_test_full").expect("request dout");
        let pd_sck_handle = chip.get_line(pd_sck_offset).expect("get pd_sck line")
            .request(LineRequestFlags::OUTPUT, 0, "hx710a_test_full").expect("request pd_sck");

        let dout   = CdevPin::new(dout_handle).expect("dout pin");
        let pd_sck = CdevPin::new(pd_sck_handle).expect("pd_sck pin");

        let connection = HX711Connection::new(dout, pd_sck);
        let mut full = Hx710aFull::new(connection).expect("init Hx710aFull");

        check_true!(full.is_ready().is_ok(), "full is_ready ok", passed, failed);

        let raw = full.read_raw();
        check_true!(raw.is_ok(), "full read_raw ok", passed, failed);
        if let Ok(v) = raw {
            check_true!(
                v >= -8_388_608 && v <= 8_388_607,
                "full read_raw in 24-bit signed range",
                passed,
                failed
            );
        }

        check_true!(full.set_rate(40).is_ok(), "set_rate 40 ok", passed, failed);
        check_true!(full.set_rate(10).is_ok(), "set_rate 10 ok", passed, failed);
        check_true!(
            matches!(full.set_rate(20), Err(HX711Error::InvalidPulseCount)),
            "set_rate 20 returns InvalidPulseCount",
            passed,
            failed
        );

        let avg = full.read_average(3);
        check_true!(avg.is_ok(), "read_average(3) ok", passed, failed);
        if let Ok(v) = avg {
            check_true!(
                v >= -8_388_608 && v <= 8_388_607,
                "read_average(3) in 24-bit signed range",
                passed,
                failed
            );
        }

        check_true!(full.tare(3).is_ok(), "tare(3) ok", passed, failed);
        let offset = full.get_offset();
        check_true!(
            offset >= -8_388_608 && offset <= 8_388_607,
            "get_offset in 24-bit signed range",
            passed,
            failed
        );

        full.set_scale(420.0);
        check_true!((full.get_scale() - 420.0).abs() < 0.001, "get_scale 420.0", passed, failed);

        let weight = full.read_weight(3);
        check_true!(weight.is_ok(), "read_weight(3) ok", passed, failed);

        let temp_raw = full.read_temperature_raw();
        check_true!(temp_raw.is_ok(), "read_temperature_raw ok", passed, failed);
        if let Ok(v) = temp_raw {
            check_true!(
                v >= -8_388_608 && v <= 8_388_607,
                "read_temperature_raw in 24-bit signed range",
                passed,
                failed
            );
        }
    }

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}
